galley/pkg/config/analysis/analyzers/injection/injection.go

```diff
@@ -99,13 +106,49 @@ func (a *Analyzer) Analyze(c analysis.Context) {
 
 		if proxyImage == "" {
 			c.Report(metadata.K8SCoreV1Pods, msg.NewPodMissingProxy(r, pod.Name, pod.GetNamespace()))
+		} else if parts := strings.Split(proxyImage, ":"); len(parts) == 2 {
+			// Note the pod/version to check later after we've collected all injector versions
+			podVersions = append(podVersions, podVersion{
+				Entry:        r,
+				ProxyVersion: parts[1]})
 		}
 
-		// TODO: if the pod is injected, check that it's using the right image. This would
-		// cover scenarios where Istio is upgraded but pods are not restarted.
-		// This is challenging because getting the expected image for the current version
-		// of Istio is non-trivial and non-standard across versions
-
 		return true
 	})
+
+	for iv := range injectorVersions {
+		for _, pv := range podVersions {
+			if pv.ProxyVersion != iv {
+				c.Report(metadata.K8SCoreV1Pods, msg.NewIstioProxyVersionMismatch(pv.Entry, pv.ProxyVersion, iv))
+			}
+		}
+	}
+}
+
+type podVersion struct {
+	Entry        *resource.Entry
+	ProxyVersion string
+}
+
+// tryReturnSidecarInjectorVersion returns an empty string if the pod is not
+// the sidecar injector; otherwise the version of the injector image is
+// returned.
+func tryReturnSidecarInjectorVersion(p *v1.Pod) string {
+	if p.Labels["app"] != "sidecarInjectorWebhook" {
+		return ""
+	}
+
+	for _, c := range p.Spec.Containers {
+		if c.Name != injectorName {
+			continue
+		}
+
+		parts := strings.Split(c.Image, ":")
+		if len(parts) != 2 {
+			continue
+		}
+		return parts[1]
```

Comment:
1. There should probably be a separate function for getting the version, it could be shared with the code in Analyze that's doing the same split.
2. Skipping over anything pods that don't have a tag on their image is likely the correct behavior, but we should be explicit about it. Add a comment explaining the intended behavior (and why it makes sense), or even better add an annotated test case.

Reply:
1. Done.

2. Added a comment.